File: week1_baseline/python/02_the_registry/lib/boukensha/registry.py
from .errors import UnknownToolError
from .tool import Tool


class Registry:
    def __init__(self, context):
        self.context = context

    def tool(self, name, *, description, parameters=None):
        def decorator(block):
            tool = Tool(name, description, parameters or {}, block)
            self.context.register_tool(tool)
            return block

        return decorator

    def dispatch(self, name, args=None):
        tool = self.context.tools.get(name)
        if not tool:
            raise UnknownToolError(f"No tool registered as '{name}'")
        return tool.block(**(args or {}))

File: week1_baseline/rust/09_global_executable/src/registry.rs
use std::collections::HashMap;

use serde_yaml_ng::Value;

use crate::context::Context;
use crate::errors::{DispatchError, ToolError, UnknownToolError};
use crate::tasks::Task;
use crate::tool::Tool;

pub struct Registry<T> {
    pub context: Context<T>,
}

impl<T: Task> Registry<T> {
    pub fn new(context: Context<T>) -> Self {
        Self { context }
    }

    pub fn tool(
        &mut self,
        name: impl Into<String>,
        description: impl Into<String>,
        parameters: Value,
        block: impl Fn(&HashMap<String, String>) -> Result<String, String> + 'static,
    ) {
        self.context.register_tool(Tool::new(name, description, parameters, block));
    }

    pub fn dispatch(&self, name: &str, args: &HashMap<String, String>) -> Result<String, DispatchError> {
        let tool = self
            .context
            .tools
            .get(name)
            .ok_or_else(|| DispatchError::UnknownTool(UnknownToolError { name: name.to_string() }))?;
        (tool.block)(args).map_err(|msg| DispatchError::ToolFailed(ToolError(msg)))
    }
}
